go/vt/wrangler/vexec_plan.go

```diff
@@ -0,0 +1,261 @@
+/*
+Copyright 2020 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package wrangler
+
+import (
+	"fmt"
+
+	"vitess.io/vitess/go/vt/sqlparser"
+)
+
+type vexecPlan struct {
+	query       string
+	opcode      int
+	parsedQuery *sqlparser.ParsedQuery
+}
+
+const (
+	insertQuery = iota
+	updateQuery
+	deleteQuery
+	selectQuery
+)
+
+func (vx *vexec) buildVExecPlan() (*vexecPlan, error) {
+	stmt, err := sqlparser.Parse(vx.query)
+	if err != nil {
+		return nil, err
+	}
+	var plan *vexecPlan
+	switch stmt := stmt.(type) {
+	case *sqlparser.Insert:
+		plan, err = vx.buildInsertPlan(stmt)
+	case *sqlparser.Update:
+		plan, err = vx.buildUpdatePlan(stmt)
+	case *sqlparser.Delete:
+		plan, err = vx.buildDeletePlan(stmt)
+	case *sqlparser.Select:
+		plan, err = vx.buildSelectPlan(stmt)
+	default:
+		return nil, fmt.Errorf("query not supported by vexec: %s", sqlparser.String(stmt))
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	plan.query = vx.query
+	vx.plan = plan
+	return plan, nil
+}
+
+func splitAndExpression(filters []sqlparser.Expr, node sqlparser.Expr) []sqlparser.Expr {
+	if node == nil {
+		return filters
+	}
+	switch node := node.(type) {
+	case *sqlparser.AndExpr:
+		filters = splitAndExpression(filters, node.Left)
+		return splitAndExpression(filters, node.Right)
+	}
+	return append(filters, node)
+}
+
+func (vx *vexec) analyzeWhere(where *sqlparser.Where) []string {
+	var cols []string
+	if where == nil {
+		return cols
+	}
+	exprs := splitAndExpression(nil, where.Expr)
+	for _, expr := range exprs {
+		switch expr := expr.(type) {
+		case *sqlparser.ComparisonExpr:
+			qualifiedName, ok := expr.Left.(*sqlparser.ColName)
+			if ok {
+				cols = append(cols, qualifiedName.Name.String())
+			}
+		}
+	}
+	return cols
+}
+
+func (vx *vexec) addDefaultWheres(where *sqlparser.Where) *sqlparser.Where {
+	cols := vx.analyzeWhere(where)
+	var hasDBName, hasWorkflow bool
+	for _, col := range cols {
+		if col == "db_name" {
+			hasDBName = true
+		} else if col == "workflow" {
+			hasWorkflow = true
+		}
+	}
+	newWhere := where
+	if !hasDBName {
+		expr := &sqlparser.ComparisonExpr{
+			Left:     &sqlparser.ColName{Name: sqlparser.NewColIdent("db_name")},
+			Operator: sqlparser.EqualStr,
+			Right:    sqlparser.NewStrVal([]byte(vx.masters[0].DbName())),
```

Comment:
This is not 100% correct, but will work for now. Each shard can actually have a different db name. But we can fix this correctly later by moving this code to `VReplicationExec`.

Reply:
understood